miner/agent.go

```diff
@@ -58,8 +61,12 @@ func (self *CpuAgent) Stop() {
 }
 
 func (self *CpuAgent) Start() {
-	self.mu.Lock()
 	defer self.mu.Unlock()
+	self.mu.Lock()
+
+	if atomic.LoadInt32(&self.isMining) == 1 {
+		return // agent already started
+	}
```

Comment:
It's not really about the performance, it's the code validity. It looks very strange that an atomic operation is protected by a mutex, and the moment that mutex disappears (somebody moves stuff around), the code will immediately become racey.